src-java/floodlight-service/floodlight-modules/src/main/java/org/openkilda/floodlight/command/rulemanager/OfBatchHolder.java

```diff
@@ -236,4 +200,111 @@ public SpeakerCommandResponse getResult() {
                 .success(failedUuids.isEmpty())
                 .failedUuids(failedUuids).build();
     }
+
+
+    private static class ExecutionGraph {
+        private Map<String, Node> nodes = new HashMap<>();
+        private boolean ready;
+        private List<String> topologicalOrder = new ArrayList<>();
+        private List<List<String>> stages = new ArrayList<>();
+        private int currentStage = 0;
+
+
+
+        public List<String> getCurrent() {
+            if (!ready) {
+                buildStages();
+            }
+            return stages.get(currentStage);
+        }
+
+        public boolean hasNext() {
+            return currentStage + 1 < stages.size();
+        }
+
+        public void next() {
+            currentStage += 1;
+        }
+
+        public List<String> getNodeInEdges(String uuid) {
+            return nodes.get(uuid).getInEdges().stream()
+                    .map(Node::getUuid).collect(Collectors.toList());
+        }
+
+        private void buildStages() {
+            topologicalSort();
+            Map<Integer, List<String>> stagesMap = new HashMap<>();
+            for (String uuid : topologicalOrder) {
```

Comment:
It must be iterated in reverse order, as far as I undesrstood (or topologicalOrder must be filled from the statrt `.addFirst()`).